src/features/DisplayControls/NursingTasks/utils/NursingTasksUtils.js

```diff
@@ -239,9 +254,125 @@ export const disableTaskTilePastNextSlotTime = (
     const upcomingTask = medicationNursingTasks[index + 1][0];
     const upcomingTaskTimeInEpoch = upcomingTask.startTimeInEpochSeconds;
     const currentTimeInEpoch = moment().unix();
-    if (upcomingTaskTimeInEpoch <= currentTimeInEpoch)
+    if (upcomingTaskTimeInEpoch <= currentTimeInEpoch) {
       currentTask.isDisabled = true;
+    }
   }
 };
 
 export const getTimeInSeconds = (days) => days * 86400;
+
+export const sortNursingTasks = (medicationNursingTasks) => {
+  medicationNursingTasks.sort((a, b) => {
+    const aTime =
+      a &&
+      (a[0]?.startTimeInEpochSeconds ??
+        a[0]?.administeredTimeInEpochSeconds ??
+        a[0]?.executionEndTime);
+    const bTime =
+      b &&
+      (b[0]?.startTimeInEpochSeconds ??
+        b[0]?.administeredTimeInEpochSeconds ??
+        b[0]?.executionEndTime);
+    return aTime - bTime;
+  });
+};
+
+export const ExtractNonMedicationTasks = (
+  nonMedicationTasks,
+  filterValue,
+  isReadMode
+) => {
+  const extractedData = [];
+  const groupedData = [],
+    completedExtractedData = [],
+    pendingExtractedData = [];
+  let currentStartTime = null;
+  let currentGroup = [];
+  nonMedicationTasks?.forEach((nonMedicationTask) => {
+    const {
+      name,
+      partOf,
+      requestedStartTime,
+      status,
+      uuid,
+      token,
+      taskType,
+      creator,
+      executionEndTime,
+    } = nonMedicationTask;
+    const startTimeInDate = new Date(requestedStartTime);
+    const taskInfo = {
+      drugName: name,
+      uuid,
+      startTimeInEpochSeconds: requestedStartTime / 1000,
+      startTime: startTimeInDate.toLocaleTimeString([], {
+        hour: "2-digit",
+        minute: "2-digit",
+        hourCycle: "h23",
+      }),
+      partOf,
+      isDisabled: isReadMode ? true : !!executionEndTime,
+      status,
+      isANonMedicationTask: true,
+      token,
+      taskType,
+      creator,
+    };
+
+    if (
+      (filterValue.id === "pending" || filterValue.id === "allTasks") &&
+      taskInfo.status === "REQUESTED"
+    ) {
+      pendingExtractedData.push(taskInfo);
+    } else if (
+      (filterValue.id === "completed" || filterValue.id === "allTasks") &&
+      taskInfo.status === "COMPLETED"
+    ) {
+      completedExtractedData.push(taskInfo);
+    }
+  });
+  extractedData.push(...pendingExtractedData);
+  extractedData.push(...completedExtractedData);
+  if (filterValue.id === "pending") {
+    extractedData.forEach((item) => {
+      if (item.partOf == null && item.taskType.display !== "nursing_activity") {
+        groupedData.push(item);
+      } else {
+        if (item.startTime !== currentStartTime && !item.stopTime) {
+          if (currentGroup.length > 0) {
+            groupedData.push(currentGroup);
+          }
+          currentGroup = [item];
+          currentStartTime = item.startTime;
+        } else {
+          currentGroup.push(item);
```

Comment:
can you add comments in the code for the logic? it would help understanding the implementation